module/zfs/arc.c

```diff
@@ -6463,11 +6467,14 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 			arc_hdr_clear_flags(hdr, ARC_FLAG_PRIO_ASYNC_READ);
 
 		/*
-		 * At this point, we have a level 1 cache miss.  Try again in
-		 * L2ARC if possible.
+		 * At this point, we have a level 1 cache miss or a blkptr
+		 * with embedded data.  Try again in L2ARC if possible.
 		 */
 		ASSERT3U(HDR_GET_LSIZE(hdr), ==, lsize);
 
+		if (embedded_bp)
+			goto zio_read;
```

Comment:
Can we add an `ASSERT3P(vd, ==, NULL)` to this conditional to make it clear that we must not have acquired the `SCL_L2ARC` lock above.  Also a brief comment explaining that the block will be read from the embedded bp and never the L2ARC.

Reply:
Sure, will do.

Reply:
Personally I would prefer to not use a `goto` for this, but I'm OK with the code as-is.

If you wanted to remove the goto, I think you'd need to wrap the DTRACE_PROBE/ARCSTAT stuff with `if (!embedded_bp)`, around lines 6485 and 6583.

Reply:
I agree with @ahrens comment. It makes more sense to be explicit about the code we want to bypass when we have an `embedded_bp`. This also ensures that we don't end up with a bug in the future because code was inadvertently bypassed by the `goto`.

Reply:
Here's the result of the obvious de-goto-ization: [arc-read-embeddedbp-next](https://github.com/dweeezil/zfs/tree/arc-read-embeddedbp-next).

Reply:
You know, due the extra level of indentation I actually like the version with the `goto` a little better, which surprises me.  Why don't we go with that version (the existing PR) unless someone feels strongly about it.

Reply:
I was originally thinking that the code change would just look like this:

Reply:
That seems perfectly reasonable and avoids the stat bump which was the main goal of this patch.  I'll re-push this in a moment.